hash/hash.go

```diff
@@ -0,0 +1,127 @@
+package hash
+
+import (
+	"bytes"
+	"crypto/subtle"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/fastrand"
+	"golang.org/x/crypto/argon2"
+)
+
+var (
+	ErrInvalidHash               = errors.New("the encoded hash is not in the correct format")
+	ErrIncompatibleVersion       = errors.New("incompatible version of argon2")
+	ErrMismatchedHashAndPassword = errors.New("passwords do not match")
+
+	// config is the configuration of the argon2id hasher.
+	// For the moment we'll use the setup we had with Kratos. Later on we can
+	// move this to a configuration file, if needed.
+	config = argon2Config{
+		SaltLength:  16,
+		Iterations:  2,
+		Memory:      131072,
+		Parallelism: 1,
+		KeyLength:   16,
+	}
+)
+
+type argon2Config struct {
+	SaltLength  uint32
+	Iterations  uint32
+	Memory      uint32
+	Parallelism uint8
+	KeyLength   uint32
+}
+
+// Generate hashes the given password with the default configuration.
+func Generate(password []byte) ([]byte, error) {
+	salt := fastrand.Bytes(int(config.SaltLength))
+	hash := argon2.IDKey(password, salt, config.Iterations, config.Memory, config.Parallelism, config.KeyLength)
+
+	var b bytes.Buffer
+	if _, err := fmt.Fprintf(
+		&b,
+		"$argon2id$v=%d$m=%d,t=%d,p=%d$%s$%s",
+		argon2.Version, config.Memory, config.Iterations, config.Parallelism,
+		base64.RawStdEncoding.EncodeToString(salt),
+		base64.RawStdEncoding.EncodeToString(hash),
+	); err != nil {
+		return nil, errors.AddContext(err, "failed to generate password hash")
+	}
+
+	return b.Bytes(), nil
+}
+
+// Compare verifies whether the given password matches the given hash.
+func Compare(password []byte, hash []byte) error {
```

Comment:
maybe something like `func(us UserSecret) Equals(other UserSecret) bool`?

Reply:
Wait... These two things are not two of a kind. One of them is a plaintext secret coming from the user and the other is the hashed version of a secret that we store in the database. That's why I've chosen the variable names there.

I was going to respond to https://github.com/SkynetLabs/skynet-accounts/pull/61#discussion_r700884476 directly but this is a very good example of the problem that would create. `UserSecret` does not tell us what we have in there - whether it's a plaintext, a ciphertext, or a hash. Or a "key" in the case of argon2, which functions like a hash for our purposes. I agree that neither does `[]byte` but that at least doesn't add to the confusion. 

I propose to keep referring to argon2 keys as hashes because we use them as such and it will add less confusion.  
I will add a type called `Argon2HashRecord` for the `[]byte` returned by argon2.  
I will change the `Compare` function to accept a string and a Argon2HashRecord types, so it's clearer what should go where.

Does that sound like a good solution?